core/src/main/scala/akka/kafka/scaladsl/Producer.scala

```diff
@@ -0,0 +1,56 @@
+/*
+ * Copyright (C) 2014 - 2016 Softwaremill <http://softwaremill.com>
+ * Copyright (C) 2016 Lightbend Inc. <http://www.lightbend.com>
+ */
+package akka.kafka.scaladsl
+
+import scala.concurrent.Future
+import akka.NotUsed
+import akka.stream.scaladsl.Flow
+import akka.stream.scaladsl.Sink
+import org.apache.kafka.clients.producer.KafkaProducer
+import org.apache.kafka.clients.producer.ProducerRecord
+import akka.kafka.internal.ProducerStage
+import akka.stream.scaladsl.Keep
+import akka.kafka.ProducerSettings
+
+object Producer {
+
+  final case class Message[K, V, +PassThrough](
+    record: ProducerRecord[K, V],
+    passThrough: PassThrough
+  )
+
+  final case class Result[K, V, PassThrough](
+    offset: Long,
+    message: Message[K, V, PassThrough]
+  )
+
+  /**
+   * Publish records to Kafka topics.
+   */
+  def plainSink[K, V](settings: ProducerSettings[K, V]): Sink[ProducerRecord[K, V], NotUsed] =
+    Flow[ProducerRecord[K, V]].map(record => Message(record, NotUsed))
+      .viaMat(flow(settings))(Keep.right)
+      .to(Sink.ignore)
+
+  /**
+   * Sink that is aware of the committable offset from a [[Consumer]]. It will commit the consumer offset
+   * when the message has been published successfully to the topic.
+   */
+  def commitableSink[K, V](settings: ProducerSettings[K, V]): Sink[Message[K, V, Consumer.Committable], NotUsed] =
+    flow[K, V, Consumer.Committable](settings)
+      .mapAsync(1)(identity)
+      .mapAsync(1)(_.message.passThrough.commit())
+      .to(Sink.ignore)
+
+  /**
+   * Publish records to Kafka topics and then continue the flow. Possibility to pass through a message, which
+   * can for example be a [[Consumer.CommittableOffset]] or [[Consumer.CommittableOffsetBatch]] that can
+   * be committed later in the flow.
+   */
+  def flow[K, V, PassThrough](settings: ProducerSettings[K, V]): Flow[Message[K, V, PassThrough], Future[Result[K, V, PassThrough]], NotUsed] =
+    Flow.fromGraph(new ProducerStage[K, V, PassThrough](() => settings.creteKafkaProducer())) //FIXME is that Future needed?
```

Comment:
@13h3r @drewhk We discussed the need for the `Future` in the output elements and had slightly different opinions. To understand better I tried this example, which is not far from reality:

    val rnd = new Random(0)
    val start = System.nanoTime()
    Source(1 to 100)
      .map { n =>
        println(s"# upstream $n")
        n
      }
      .map { n =>
        Future {
          println(s"# produce begin $n")
          Thread.sleep(100 + (rnd.nextDouble() * 100).toLong)
          println(s"# produce end   $n")
          n
        }
      }
      .mapAsync(1) { n =>
        n
      }
      .map { n =>
        println(s"# downstream $n, ${(System.nanoTime() - start).nanos.toMillis} ms")
        n
      }
      .runWith(Sink.ignore)

    Thread.sleep(20000)

The results for completing the 100 elements with `withAutoFusing(true)`:

* mapAsync(1): 15804 ms
* mapAsync(10): 1877 ms
* mapAsyncUnordered(10): 1670 ms

and the results with `withAutoFusing(false)`:

* mapAsync(1): 1340 ms
* mapAsync(10): 960 ms
* mapAsyncUnordered(10): 823 ms

I don't think users will get this, and we should remove the Future and define a parallelism setting.

WDYT?

It's also important that an earlier commit doesn't overwrites a later commit. That could happen in my example because of the spawned Future, but in the real code I think it's safe because the calls are serialized by the stage and then I assume that the kafka driver will execute things sequentially.